File: Linked_List_free_code_camp/4.linked_list_find.js
class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

const head =
    new Node(1,
        new Node(2,
            new Node(3,
                new Node(4,
                    new Node(8, null)))));

// A -> B -> C -> D -> null

const linkedListFind = (head, target) => {
    let ans = false;
    let curr = head;
    while (curr) {
        if (curr.val === target) {
            ans = true;
            break;
        }
        curr = curr.next;
    }
    return ans;
};

console.log(linkedListFind(head, 3));


const linkedListFindRecursive = (head, target) => {
    if (head == null) return false;

    const faith = linkedListFindRecursive(head.next, target);

    return head.val === target || faith;
};

console.log(linkedListFindRecursive(head, 5));

File: Linked_List_free_code_camp/4.linked_list_find.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;

    Node(int val, Node* next = NULL) {
        this->val = val;
        this->next = next;
    }
};

void print_v(vector<int> values) {
    for (auto e : values) {
        cout << e << " ";
    }
    cout << endl;
}

bool linkedListFindIterative(Node* head, int target) {
    Node* t = head;

    while (t) {
        if (t->val == target) {
            return true;
        }
        t = t->next;
    }

    return false;
};

bool linkedListFindRecursive(Node* head, int target) {
    if (head == NULL) return false;

    bool faith = linkedListFindRecursive(head->next, target);

    return head->val == target || faith;
};


int main() {

    Node* head =
        new Node(10,
            new Node(20,
                new Node(30,
                    new Node(40))));

    cout << linkedListFindIterative(head, 20) << endl;
    cout << linkedListFindIterative(head, 80) << endl;
    cout << linkedListFindRecursive(head, 20) << endl;
    cout << linkedListFindRecursive(head, 80) << endl;

    return 0;
}